File: examples/default-args-pitfalls/default-args.cs
using System;
using System.Collections.Generic;

public class DefaultArgs
{
    public static List<int> CollectNumbers(int n, List<int> result = null)
    {
        if (result == null) result = new List<int>();
        
        if (n <= 0)
            return result;
        
        result.Add(n);
        return CollectNumbers(n - 1, result);
    }
    
    public static void Main()
    {
        Console.WriteLine("First call: [" + string.Join(", ", CollectNumbers(3)) + "]");
        Console.WriteLine("Second call: [" + string.Join(", ", CollectNumbers(2)) + "]");
        Console.WriteLine("Third call: [" + string.Join(", ", CollectNumbers(4)) + "]");
    }
}

File: examples/default-args-pitfalls/default-args.go
package main

import "fmt"

func collectNumbers(n int, result []int) []int {
    if result == nil {
        result = []int{}
    }
    
    if n <= 0 {
        return result
    }
    
    result = append(result, n)
    return collectNumbers(n-1, result)
}

func main() {
    fmt.Println("First call:", collectNumbers(3, nil))
    fmt.Println("Second call:", collectNumbers(2, nil))
    fmt.Println("Third call:", collectNumbers(4, nil))
}